File: impl/tree/segment.go
package tree

import (
	"log"
)

type Comp func(interface{}, interface{}) interface{}
type Mutate func(interface{}) interface{}

type SegmentTree struct {
	tree []interface{}
	lazy [][]Mutate
	comp Comp
	len  int
}

func NewSegmentTree(arr []interface{}, comp Comp) *SegmentTree {
	backing := 1
	for backing < len(arr) {
		backing <<= 1
	}

	tree := &SegmentTree{
		make([]interface{}, 2*backing-1),
		make([][]Mutate, 2*backing-1),
		comp,
		len(arr),
	}
	tree.build(arr, 0, 0, tree.len-1)

	return tree
}

func (seg *SegmentTree) build(arr []interface{}, node, start, end int) {
	if start == end {
		seg.tree[node] = arr[start]
		return
	}

	mid := (start + end) / 2
	seg.build(arr, 2*node+1, start, mid)
	seg.build(arr, 2*node+2, mid+1, end)
	seg.tree[node] = seg.comp(seg.tree[2*node+1], seg.tree[2*node+2])
}

func (seg *SegmentTree) Print() {
	seg.print(0, 0, seg.len-1)
}

func (seg *SegmentTree) print(node, start, end int) {
	log.Printf(
		"Node: %d Value: %d Range: %d-%d Lazy: %d\n",
		node, seg.tree[node], start, end, len(seg.lazy[node]),
	)

	if start == end {
		return
	}

	mid := (start + end) / 2
	seg.print(2*node+1, start, mid)
	seg.print(2*node+2, mid+1, end)
}

func (seg *SegmentTree) propagate(node int, mut Mutate, leaf bool) {
	seg.tree[node] = mut(seg.tree[node])

	if !leaf {
		seg.lazy[2*node+1] = append(seg.lazy[2*node+1], mut)
		seg.lazy[2*node+2] = append(seg.lazy[2*node+2], mut)
	}
}

func (seg *SegmentTree) Update(start, end int, mut Mutate) {
	seg.update(0, 0, seg.len-1, start, end, mut)
}

func (seg *SegmentTree) update(node, left, right, start, end int, mut Mutate) {
	for _, lazy := range seg.lazy[node] {
		seg.propagate(node, lazy, left == right)
	}
	seg.lazy[node] = nil

	if start > right || end < left {
		return
	}

	if left >= start && right <= end {
		seg.propagate(node, mut, left == right)
		return
	}

	mid := (left + right) / 2
	seg.update(2*node+1, left, mid, start, end, mut)
	seg.update(2*node+2, mid+1, right, start, end, mut)
	seg.tree[node] = seg.comp(seg.tree[2*node+1], seg.tree[2*node+2])
}

func (seg *SegmentTree) Query(start, end int) interface{} {
	return seg.query(0, 0, seg.len-1, start, end)
}

func (seg *SegmentTree) query(node, left, right, start, end int) interface{} {
	if start <= left && end >= right {
		return seg.tree[node]
	} else if start > right || end < left {
		return nil
	}

	mid := (left + right) / 2
	l := seg.query(2*node+1, left, mid, start, end)
	r := seg.query(2*node+2, mid+1, right, start, end)

	if r == nil {
		return l
	} else if l == nil {
		return r
	}

	return seg.comp(l, r)
}

File: src/trees/segment.rs
macro_rules! left {
    ($node:expr) => {
        2 * $node + 1
    };
}

macro_rules! right {
    ($node:expr) => {
        2 * $node + 2
    };
}

#[derive(Debug)]
pub struct SegmentTree<T: Copy + Default> {
    tree: Vec<T>,
    comb: fn(T, T) -> T,
    len: usize,
}

impl<T: Copy + Default> SegmentTree<T> {
    pub fn new(arr: Vec<T>, comb: fn(T, T) -> T) -> SegmentTree<T> {
        let mut backing = 1;
        while backing < arr.len() {
            backing <<= 1;
        }
        backing = 2 * backing - 1;

        let mut s = SegmentTree {
            tree: vec![Default::default(); backing],
            comb,
            len: arr.len() as usize,
        };
        s.build(&arr, 0, 0, s.len - 1);
        s
    }

    fn build(&mut self, arr: &Vec<T>, node: usize, start: usize, end: usize) {
        if start == end {
            self.tree[node] = arr[start];
            return;
        }

        let mid = (start + end) / 2;
        self.build(arr, left!(node), start, mid);
        self.build(arr, right!(node), mid + 1, end);
        self.tree[node] = (self.comb)(self.tree[left!(node)], self.tree[right!(node)]);
    }

    fn range(&self, start: usize, end: usize) -> Option<T> {
        self.query_recurse(0, 0, self.len - 1, start, end)
    }

    fn query_recurse(
        &self,
        node: usize,
        left: usize,
        right: usize,
        start: usize,
        end: usize,
    ) -> Option<T> {
        if start <= left && end >= right {
            return Some(self.tree[node]);
        } else if start > right || end < left {
            return None;
        }

        let mid = (left + right) / 2;
        let l = self.query_recurse(left!(node), left, mid, start, end);
        let r = self.query_recurse(right!(node), mid + 1, right, start, end);

        match (l, r) {
            (_, None) => l,
            (None, _) => r,
            (Some(lv), Some(rv)) => Some((self.comb)(lv, rv)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn segment_tree() {
        let arr = vec![1, 3, 5, 7, 9, 11];
        println!("Array: {:#?}", arr);

        let tree = SegmentTree::new(arr, |a, b| a + b);
        println!("Tree: {:#?}", tree);

        if let Some(sum) = tree.range(0, 5) {
            println!("{}", sum);
        }
    }
}
